src/mmw/apps/geoprocessing_api/calcs.py

```diff
@@ -50,6 +50,70 @@ def animal_population(geojson):
     }
 
 
+def stream_data(geojson):
+    """
+    Given a GeoJSON shape, retreive stream data from the `nhdflowline` table
+    to display in the Analyze tab
+
+    Returns a dictionary to append to outgoing JSON for analysis results.
+    """
+
+    geom = GEOSGeometry(geojson, srid=4326)
+    table_name = 'nhdflowline'
+
+    sql = '''
+        SELECT lengthkm, stream_order
+        FROM {table_name}
+        WHERE ST_Intersects(geom, ST_SetSRID(ST_GeomFromText(%s), 4326))
+        '''.format(table_name=table_name)
+
+    with connection.cursor() as cursor:
+        cursor.execute(sql, [geom.wkt])
+
+        if cursor.rowcount:
+            columns = [col[0] for col in cursor.description]
+            streams = [
+                dict(zip(columns,
+                         [
+                             float(row[0]) if row[0] else 0,
+                             int(row[1]) if row[1] and row[1] != 0 else 999,
+                         ]))
+                for row in cursor.fetchall()
+            ]
+        else:
+            streams = []
+
+    stream_data = {
+        str(x): {
+            "stream_order": x,
+            "lengthkm": 0,
+            "slopepct": 0,
+            "ag_length": 0,
+            "non_ag_length": 0,
+        } for x in list(range(1, 11)) + [999]
+    }
+
+    for s in streams:
+        stream_data[str(s['stream_order'])]['lengthkm'] += s['lengthkm']
```

Comment:
Instead of doing this aggregation in Python, it'll be much faster to do it in the query itself:

```sql
SELECT SUM(lengthkm) AS total_lengthkm, stream_order
FROM nhdflowline
WHERE ST_Intersects(geom, ST_SetSRID(ST_GeomFromGeoJSON(%s), 4326))
GROUP BY stream_order;
```

Reply:
7f07e18 adds this change; since the data returned from the sql query doesn't always include a full set of all stream orders, also added a loop to use `setdefault` to merge in 0 values for missing stream orders.